src/mongocxx/options/index.hpp

```diff
@@ -0,0 +1,423 @@
+        const int type() const override;
+        stdx::optional<std::string> _config_str;
+    };
+
+    index();
+
+    ///
+    /// Whether or not to build the index in the background so that building
+    /// the index does not block other database activities.
+    ///
+    /// @param background
+    ///   Whether or not to build the index in the background.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    void background(bool background);
+
+    ///
+    /// The current background setting.
+    ///
+    /// @return The current background.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    const stdx::optional<bool>& background() const;
+
+    ///
+    /// Whether or not to create a unique index so that the collection will not
+    /// accept insertion of documents where the index key or keys match an
+    /// existing value in the index.
+    ///
+    /// @param unique
+    ///   Whether or not to create a unique index.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    void unique(bool unique);
+
+    ///
+    /// The current unique setting.
+    ///
+    /// @return The current unique.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    const stdx::optional<bool>& unique() const;
+
+    ///
+    /// The name of the index.
+    ///
+    /// @param name
+    ///   The name of the index.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    void name(std::string name);
+
+    ///
+    /// The current name setting.
+    ///
+    /// @return The current name.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    const stdx::optional<std::string>& name() const;
+
+    ///
+    /// Whether or not to create a sparse index. Sparse indexes only reference
+    /// documents with the indexed fields.
+    ///
+    /// @param sparse
+    ///   Whether or not to create a sparse index.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    void sparse(bool sparse);
+
+    ///
+    /// The current sparse setting.
+    ///
+    /// @return The current sparse.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    const stdx::optional<bool>& sparse() const;
+
+    ///
+    /// Optionally used only in MongoDB 3.0.0 and higher. Specifies the storage
+    /// engine options for the index
+    ///
+    /// @param storage_options
+    ///   The storage engine options for the index in.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+    ///
+    /// @warning
```

Comment:
I'd move this warning up higher, to right after the description, before the @param part.